numpy/core/src/multiarray/mapping.c

```diff
@@ -252,7 +252,26 @@ unpack_indices(PyObject *index, PyObject **result, npy_intp result_n)
 }
 
 /**
- * Prepare an npy_index_object from the python slicing object.
+ * If arg is a callable, apply it to self.
+ *
+ * @returns arg, arg(self), or NULL. NULL is returned if arg(self) fails
+ *      (if the callable takes more than a single argument, for example).
+ */
+static NPY_INLINE PyObject *
+maybe_apply_callable(PyArrayObject *self, PyObject *arg)
+{
+    if (PyFunction_Check(arg)) {
+        // NOTE: Starting with Python 3.9, we can instead do
+        //   PyObject *result = (PyObject *) PyObject_CallOneArg(arg, self);
+        PyObject *self_as_tuple = PyTuple_Pack(1, self);
+        PyObject *result = (PyObject *) PyObject_CallObject(arg, self_as_tuple);
+        return (result == NULL) ? NULL : result;
```

Comment:
My reading of the docs is that this leaks a reference
```suggestion
        PyObject *self_as_tuple = PyTuple_Pack(1, self);
        PyObject *result = (PyObject *) PyObject_CallObject(arg, self_as_tuple);
        Py_DECREF(self_as_tuple);
        return result;
```
Alternatively, you could use
```suggestion
        return (PyObject *) PyObject_CallFunctionObjArgs(arg, self, NULL);
```
I don't know which is more performant

Reply:
```python
In [1]: import numpy as np; arr = np.arange(4)

In [2]: %timeit arr[lambda x: x > 1]
```

first suggestion: 760 ns ± 0.614 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
second suggestion: 760 ns ± 2.47 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

No difference, and that second suggestions is a very clean one-liner